### PR TITLE
Add typing to otfautohint

### PR DESCRIPTION
**Pull request by simoncozens**
## Description

otfautohint is a complex piece of code, and it's easy to get lost; it's even easier to make mistakes. Having typing information available makes it easier to catch such mistakes, and allows editors with Python type inferencing (Visual Code with pylance, Sublime Text with pyright, etc.) to provide type hints for variables and methods. This in turn helps to find typos, dead code, and so on.

## Checklist:

- [X] I have followed the [Contribution Guidelines](https://github.com/adobe-type-tools/afdko/blob/develop/CONTRIBUTING.md)
- [ ] ~I have added **test code and data** to prove that my code functions correctly~ (N/A)
- [X] I have verified that new and existing tests pass locally with my changes
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~ (N/A)

### CODE REVIEW

**Comment by skef** [comment]
As you're probably realizing this has a lot of formatting related problems (flake8), some of which are due to edits for line length that you've removed.

**Comment by josh-hadley** [comment]
@simoncozens thanks for this contribution. It looks like there are a number of linter failures that are blocking the tests from running. I recommend that you run `flake8` using the config from this repo, fix all reported issues, and resubmit. Once we get the linter + tests green we will have a closer review and see about getting it merged.

**Comment by simoncozens** [comment]
Yeah, sorry, I was assuming the code formatting was black/PEP8 so my editor runs black on save to keep things tidy. I'll try and put things back into your preferred style.

**Comment by skef** [comment]
I think I would prefer a submission with the _substantive_ changes and then (perhaps) a different submission with the large amount of _stylistic_ changes. Or perhaps just skip the latter? Best to either mesh one's contributions with the existing style or talk in advance about what you want to change and why.

**Comment by skef** [comment]
@josh-hadley Merging this would presumably require additional `pytype` github checks, so that notations like `# pytype: disable=attribute-error` won't just rot.

**Comment by simoncozens** [comment]
To be honest, it doesn't currently work with pytype. I started working with pytype, and as I added annotations I found myself in situations where a method returned an instance of its own class, which needs the `Self` type, and pytype doesn't currently support that. So I switched to mypy. In a couple of cases mypy doesn't like the Self return value, but it does a better job than pytype (and continues running, whereas pytype just stops and refuses to go on).

**Comment by simoncozens** [comment]
Coverage tests are sad because apparently you can't use `weakref.ReferenceType` as a generic in Python 3.8.

File: python/afdko/otfautohint/hintstate.py [hunk 1]
```diff
@@ -369,23 +433,29 @@ def remExtraBends(self):
...
+                        hsi.type == hintSegment.sType.BEND
+                        and hsd.type != hintSegment.sType.BEND
+                        and hsd.type != hintSegment.sType.GHOST
+                        and (hsd.max - hsd.min) > (hsi.max - hsi.min) * 3
+                    ):
                         hsi.deleted = True
-                        log.debug("rem seg loc %g from %g to %g" %
-                                  (hsi.loc, hsi.min, hsi.max))
+                        log.debug(
+                            "rem seg loc %g from %g to %g" % (hsi.loc, hsi.min, hsi.max)
```
Comment (skef) [inline python/afdko/otfautohint/hintstate.py:444]: Surprised there aren't flake8 complaints about these long lines.

We're not going to go through and insist that every otherwise-unchanged line is formatted back to what it was, but let's revert the ones around here for clarity and flake8 purposes. Best to keep this PR mostly tracking what needed to be changed for these type changes.

File: python/afdko/otfautohint/hinter.py [hunk 2]
```diff
@@ -428,7 +462,7 @@ def handleOverlap(self):
         else:
             return self.isMulti
 
-    def addSegment(self, fr, to, loc, pe1, pe2, typ, desc, mid=False):
+    def addSegment(self, fr, to, loc, pe1: Optional[pathElement], pe2: Optional[pathElement], typ, desc, mid=False) -> None:
```
Comment (skef) [inline python/afdko/otfautohint/hinter.py:465]: This has flake8 problems too, doesn't it?

File: python/afdko/otfautohint/hinter.py [hunk 3]
```diff
@@ -459,29 +493,30 @@ def addSegment(self, fr, to, loc, pe1, pe2, typ, desc, mid=False):
...
         self.hs.addSegment(fr, to, loc, pe1, pe2, typ, self.Bonus,
                            self.isV(), mid1, mid2, desc)
 
-    def CPFrom(self, cp2, cp3):
+    def CPFrom(self, cp2, cp3) -> Any:
         """Return point cp3 adjusted relative to cp2 by CPFrac"""
         return (cp3 - cp2) * (1.0 - self.CPfrac) + cp2
 
-    def CPTo(self, cp0, cp1):
+    def CPTo(self, cp0, cp1) -> Any:
```
Comment (skef) [inline python/afdko/otfautohint/hinter.py:504]: What is the purpose of the Any return type notations? Does it help compared with just leaving them off?

File: python/afdko/otfautohint/hinter.py [hunk 4]
```diff
@@ -1848,14 +1900,20 @@ def calcInstanceStems(self, glidx):
 
                     iSS = iSSl[sidx][ul]
                     found = False
+                    assert self.glyph
                     if seg0.isBBox():
                         if seg0.isGBBox():
                             pbs = self.glyph.getBounds(None)
                         else:
+                            # I don't believe this can be reached, because
```
Comment (skef) [inline python/afdko/otfautohint/hinter.py:1908]: This is probably just so obscure a corner case that none of our tests exercise it and no one has hit it yet, so we didn't catch the bug. I'll take a look at some point. 

`seg0.isBBox() and not seg0.isGBBox()` implies that the stems for seg0 were generated from the bounding box of the segment itself because nothing else worked. Only very unusual splines will have that problem.

**Review by skef — COMMENTED** [review]
Some preliminary review feedback

**Comment by skef** [comment]
One reason we ported `psautohint` to python was the hope that it could be modified by the broader community, in comparison to the C code which was only ever worked on by a small group of Adobe-internal folks while squinting hard and crossing their fingers. 

This PR is both an external contribution and an effort to make further contributions easier, so in that sense it's what we want and were hoping for. 

Still, there is a basic concern about keeping these annotations from rotting. Some users will have environments that will make good use of the typing. Others will just be editing the files and running the tests and may have no idea what this stuff is or, if they do, how to change the annotations as the code changes. 

Obviously some sort of verifier that can be run as part of the GitHub checks would be ideal, but it sounds like `pytype` can't do that and `mypy` might or might not be able to. 

Do you have thoughts on this? Basically, how to maintain types in Python assuming a wide variety of users with different levels of experience and tooling? 

Also: Keep in mind that we haven't been doing this internally, so it make take a while just to digest the implications of typed Python on our end.

**Comment by anthrotype** [comment]
FWIW annotations are pretty useless unless they are actually checked by a typechecker

**Comment by anthrotype** [comment]
ok, well.. self-documentation sure, but they then tend to rot as @skef noted.

File: python/afdko/otfautohint/glyphData.py [hunk 5]
```diff
@@ -224,14 +238,14 @@ def __mul__(self, other):
         Returns a new pt object with this object's coordinates multiplied by
         a scalar value
         """
-        if not isinstance(other, numbers.Number):
+        if not isinstance(other, (int, float)):
```
[the remark this answers is not in the document]
Comment (simoncozens) [reply python/afdko/otfautohint/glyphData.py:241]: `numbers.Number` includes `complex`, which I think you should probably exclude.

File: python/afdko/otfautohint/hintstate.py [hunk 6]
```diff
@@ -136,23 +174,25 @@ def __init__(self, lloc, uloc, val, spc, lseg, useg, isGhost=False):
...
         self.lseg = lseg
         self.useg = useg
-        self.best = None
+        self.best: Optional[stemValue] = None
         self.initialVal = val
-        self.idx = None
+        self.idx: Optional[int] = None
 
-    def __eq__(self, other):
+    def __eq__(self, other: object) -> bool:
```
[the remark this answers is not in the document]
Comment (simoncozens) [reply python/afdko/otfautohint/hintstate.py:181]: Yes, apparently the `__eq__` method is [supposed to be more tolerant](https://stackoverflow.com/questions/54801832/mypy-eq-incompatible-with-supertype-object). mypy pulls a fit and complains about Liskov substitution if you make it `Self`.

**Comment by simoncozens** [comment]
> Obviously some sort of verifier that can be run as part of the GitHub checks would be ideal, but it sounds like pytype can't do that and mypy might or might not be able to.

We can certainly get mypy set up in the CI, and add a disable comment to turn off the Self checks.

File: python/afdko/otfautohint/hinter.py (hunk 3 above)
Comment (simoncozens) [reply python/afdko/otfautohint/hinter.py:504]: Python typing is progressive, and basically, I haven't finished adding all the annotations. But if you add *any* type notation, a method gets checked as best it can be, but with no type notation at all you don't get any checking or inferencing.

File: python/afdko/otfautohint/otfFont.py [hunk 7]
```diff
@@ -391,8 +399,11 @@ def getPrivateDictVal(self, pDict, attr, default, vsindex, vsi):
 
         return val
 
-    def getPrivateFDDict(self, allowNoBlues, noFlex, vCounterGlyphs,
-                         hCounterGlyphs, desc, fdIndex=0, gl_vsindex=None):
+    def getPrivateFDDict(self, allowNoBlues: bool, noFlex: bool,
+                         vCounterGlyphs, hCounterGlyphs, desc: str,
+                         fdIndex=0, gl_vsindex: Optional[int] = None) -> Union[
+        Tuple[List[fdTools.FDDict], int],  # VF
```
[the remark this answers is not in the document]
Comment (simoncozens) [reply python/afdko/otfautohint/otfFont.py:405]: Hmm. That's not very clear in the code. At the bottom of the function we have

```
            if not self.is_vf:
                return fdDict

            fdDictArray.append(fdDict)

        assert self.is_vf and gl_vsindex is not None
        return fdDictArray, gl_vsindex
```

So it looks very much like: if not `is_vf` -> single FDDict, if `is_vf` -> list and integer tuple.

[the remark this answers is not in the document]
Comment (skef) [reply python/afdko/otfautohint/otfFont.py:405]: Oh, yes, you're right, I was misremembering what this code does. The array is to return private dicts for each of the "masters" for the glyph in the font. I'll resolve.

File: python/afdko/otfautohint/glyphData.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (skef) [reply python/afdko/otfautohint/glyphData.py:241]: Fair enough.

File: python/afdko/otfautohint/hintstate.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (skef) [reply python/afdko/otfautohint/hintstate.py:183]: Good to know

**Comment by skef** [comment]
A version of this is being merged into the addfeatures branch. Closing this PR.
